src/uct/ib/mlx5/ib_mlx5_dv.c

```diff
@@ -70,10 +71,55 @@ static ucs_status_t uct_ib_mlx5_device_init(uct_ib_device_t *dev)
     return status;
 }
 
+static ucs_status_t uct_ib_mlx5_device_init(uct_ib_device_t *dev)
+{
+    ucs_status_t status = UCS_OK;
+    int ret = -1;
+
+#if HAVE_DECL_MLX5DV_DEVX_GENERAL_CMD
+    uint32_t out[UCT_ST_SZ_DW(query_hca_cap_out)] = {0};
+    uint32_t in[UCT_ST_SZ_DW(query_hca_cap_in)] = {0};
+    struct ibv_context *ctx = dev->ibv_context;
+
+    UCT_SET(query_hca_cap_in, in, opcode, MLX5_CMD_OP_QUERY_HCA_CAP);
+    UCT_SET(query_hca_cap_in, in, op_mod, HCA_CAP_OPMOD_GET_MAX |
+                                          (MLX5_CAP_GENERAL << 1));
+    ret = mlx5dv_devx_general_cmd(ctx, in, sizeof(in), out, sizeof(out));
+    if (ret == 0) {
+        if (UCT_GET(query_hca_cap_out, out, capability.cmd_hca_cap.dct)) {
+            dev->flags |= UCT_IB_DEVICE_FLAG_DC;
+        }
+        if (UCT_GET(query_hca_cap_out, out,
+                    capability.cmd_hca_cap.compact_address_vector)) {
+            dev->flags |= UCT_IB_DEVICE_FLAG_AV;
+        }
+    } else if (errno != EPERM &&
+               errno != EPROTONOSUPPORT &&
+               errno != EOPNOTSUPP) {
+        ucs_error("MLX5_CMD_OP_QUERY_HCA_CAP failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+#endif
+    if (ret != 0) {
+        status = uct_ib_mlx5_check_dc(dev);
+    }
+
+    return status;
+}
+
 UCT_IB_DEVICE_INIT(uct_ib_mlx5_device_init);
 
 #endif
 
+#if HAVE_DECL_MLX5DV_DEVX_GENERAL_CMD
+ucs_status_t uct_ib_mlx5_get_compact_av(uct_ib_iface_t *iface, int *compact_av)
```

Comment:
maybe remove this function altogether, and check UCT_IB_DEVICE_FLAG_AV during non-dv device init as well (the old way)?
also need to rename uct_ib_mlx5_device_init() to uct_ib_mlx5dv_device_init()